HttpRequest Params or indexer throws InvalidOperationException when request has content type other than form.

sdekock opened the pull request:
**PR Title**
HttpRequest Params or indexer throws InvalidOperationException when request has content type other than form.

**PR Description**
In the Params or indexer on HttpRequest, do not try to access Form property when request has content type other than form, because this throws an InvalidOperationException

Fixes #309

twsouthwick commented:
Thanks @sdekock! This looks great and thanks for adding the tests.

Should this check also be on the `Form` indexer? Do we have similar behavior if the content type is not form?

sdekock commented:
# Test results

## ASP.NET

```
GET / HTTP/1.1
```

`Request.Files` is empty `HttpFileCollection` or `HttpFileCollectionWrapper`
`Request.Form` is empty `HttpValueCollection`
`Request.Form["test"]` is null
`Request.Params["test"]` is null
`Request["test"]` is null

```
POST /?test=a HTTP/1.1
Content-Type: application/json

{ test: 'b' }
```

`Request.Files` is empty `HttpFileCollection` or `HttpFileCollectionWrapper`
`Request.Form` is empty `HttpValueCollection`
`Request.Form["test"]` is null
`Request.Params["test"]` is `"a"`
`Request["test"]` is `"a"`

```
POST /?test=a HTTP/1.1
Content-Type: application/x-www-form-urlencoded

test=b
```

`Request.Files` is empty `HttpFileCollection` or `HttpFileCollectionWrapper`
`Request.Form` is `HttpValueCollection` with entry `test` = `"b"`
`Request.Form["test"]` is `"b"`
`Request.Params["test"]` is `"a,b"`
`Request["test"]` is `"a"`

## ASP.NET Core (without changes in this PR)

```
GET / HTTP/1.1
```

**`Request.Files` throws InvalidOperationException: 'Incorrect Content-Type: '
`Request.Form` throws InvalidOperationException: 'Incorrect Content-Type: '
`Request.Form["test"]` throws InvalidOperationException: 'Incorrect Content-Type: '
`Request.Params["test"]` throws InvalidOperationException: 'Incorrect Content-Type: '
`Request["test"]` throws InvalidOperationException: 'Incorrect Content-Type: '**

```
POST /?test=a HTTP/1.1
Content-Type: application/json

{ test: 'b' }
```

**`Request.Files` throws InvalidOperationException: 'Incorrect Content-Type: application/json'
`Request.Form` throws InvalidOperationException: 'Incorrect Content-Type: application/json'
`Request.Form["test"]` throws InvalidOperationException: 'Incorrect Content-Type: application/json'**
`Request.Params["test"]` is `"a"`
`Request["test"]` is `"a"`

```
POST /?test=a HTTP/1.1
Content-Type: application/x-www-form-urlencoded

test=b
```

`Request.Files` is empty `HttpFileCollection` or `HttpFileCollectionWrapper`
`Request.Form` is `StringValuesReadOnlyDictionaryNameValueCollection` with entry `test` = `"b"`
`Request.Form["test"]` is `"b"`
**`Request.Params["test"]` is `"a"`**
`Request["test"]` is `"a"`

sdekock commented:
@twsouthwick I added the content-type check to the Files and Form properties and I have added the missing Form property to HttpRequestBase/Wrapper.

The only incompatibility still now is that Params apparently returns a combined collection of QueryString, Form, Cookies, and ServerVariables items. Unlike the indexer, which returns the first item found from those collections.
I'm unsure how I can accomplish that in an efficient way.

twsouthwick commented:
> The only incompatibility still now is that Params apparently returns a combined collection of QueryString, Form, Cookies, and ServerVariables items. Unlike the indexer, which returns the first item found from those collections.
I'm unsure how I can accomplish that in an efficient way.

Maybe we'll need to do what was done on framework. We tried to be efficient without copying things, but for this behavior we may need to. At least it won't be *worse* than framework :) See https://referencesource.microsoft.com/#System.Web/HttpRequest.cs,794 for how that was done.

sdekock commented on src/Microsoft.AspNetCore.SystemWebAdapters/HttpFileCollection.cs:
```diff
@@ -12,6 +12,8 @@ namespace System.Web;
 [SuppressMessage("Design", "CA1010:Collections should implement generic interface", Justification = Constants.ApiFromAspNet)]
 public sealed class HttpFileCollection : NameObjectCollectionBase
 {
+    internal static readonly HttpFileCollection Empty = new(new FormFileCollection());
```
The `FormFileCollection` itself is not, it derives from `List<IFormFile>`. But `HttpFileCollection` provides no way to change the contents.

I could add an empty implementation of `IFormFileCollection` ? I found a version of it already in ASP.NET Core's FormCollectionModelBinder.

twsouthwick replied:
I just took a look and line 23 is setting it to read only, so it should be good

twsouthwick commented:
Do you want to update the params for the indexer in this PR? Otherwise, it LGTM and we can merge it in.

twsouthwick commented:
Ok - seems like we should create an issue to track it and not block this fix from going in.

twsouthwick commented:
Thanks @sdekock for the fix here. I created #312 to track the issue with params - if you come up with a good solution there happy to take another PR :)
